src/hotspot/share/opto/loopopts.cpp

```diff
@@ -4120,3 +4121,204 @@ bool PhaseIdealLoop::duplicate_loop_backedge(IdealLoopTree *loop, Node_List &old
+        first_ur = current;
+        current = nullptr;
+        break; // Success.
+      } else {
+        DEBUG_ONLY( current->dump(1); )
+        assert(false, "scalar_input is neither phi nor a matchin reduction");
+        break; // Chain traversal fails.
+      }
+    }
+    if (current != nullptr) {
+      // Chain traversal was not successful.
+      continue;
+    }
+    assert(first_ur != nullptr, "must have successfully terminated chain traversal");
+
+    // Determine types
+    const TypeVect* vec_t = last_ur->vect_type();
+    uint vector_length    = vec_t->length();
+    BasicType bt          = vec_t->element_basic_type();
+    const Type* bt_t      = Type::get_const_basic_type(bt);
+
+    // Create vector of identity elements (zero for add, one for mul, etc)
+    Node* identity_scalar = ReductionNode::make_identity_input_for_reduction_from_vector_opc(_igvn, last_ur->Opcode(), bt);
+    set_ctrl(identity_scalar, C->root());
+    VectorNode* identity_vector = VectorNode::scalar2vector(identity_scalar, vector_length, bt_t);
+    _igvn.register_new_node_with_optimizer(identity_vector);
+    set_ctrl(identity_vector, C->root());
+    assert(vec_t == identity_vector->vect_type(), "matching vector type");
+#ifdef ASSERT
+    if (TraceNewVectors) {
+      tty->print("new Vector node: ");
+      identity_vector->dump();
+    }
+#endif
+
+    // Traverse down the chain of UnorderedReductions, and create the vector_accumulators.
+    current = first_ur;
+    Node* last_vector_accumulator = phi;
+    while (true) {
+      // Create vector_accumulator to replace current.
+      Node* vector_input = current->in(2);
+      VectorNode* vector_accumulator = current->make_normal_vector_op(last_vector_accumulator, vector_input, vec_t);
+      _igvn.register_new_node_with_optimizer(vector_accumulator);
+      C->copy_node_notes_to(vector_accumulator, current);
+      set_ctrl(vector_accumulator, cl);
+      last_vector_accumulator = vector_accumulator;
+      if (current != last_ur) {
+        // All UnorderedReductions except the last are now useless.
+        _igvn.rehash_node_delayed(current);
+        current->set_req_X(1, C->top(), &_igvn);
+        current->set_req_X(2, C->top(), &_igvn);
+      }
+#ifdef ASSERT
+      if (TraceNewVectors) {
+        tty->print("new Vector node: ");
+        vector_accumulator->dump();
+      }
+#endif
+      // Iterate down, until we hit last_ur.
+      if (current != last_ur) {
+        current = current->unique_out()->as_UnorderedReduction();
+      } else {
+        current = nullptr;
+        break;
+      }
+    }
+    assert(current == nullptr, "terminated correctly");
+
+    // After the loop, we can reduce the init and last_vector_accumulator.
+    Node* init = phi->in(1);
+    _igvn.rehash_node_delayed(last_ur);
+    last_ur->set_req_X(1, init, &_igvn);
+    last_ur->set_req_X(2, last_vector_accumulator, &_igvn);
+
+    // Turn the scalar phi into a vector phi.
+    _igvn.rehash_node_delayed(phi);
```

Comment:
Is it possible to setup the vector phi first, and then replace all reduction nodes by vector accumulators via calling `_igvn.replace_node()`? Current code here looks a bit wordy.

Reply:
If I set up the phi first, then where do I keep the `init` value? I have to make sure it has a use throughout the transformation, other wise it may think it is `dead`.

Reply:
Why can the `init` node be dead? I think it should be ok as long as `init` node is connected well at the end of this transformation function. Registering new nodes just puts nodes into igvn worklist and no actual igvn is performed in the middle.